### PR TITLE
Fix _has then chain then _has searches

### PR DESCRIPTION
**Pull request by lukedegruchy**
- Fix the Has predicate to properly handle a chain within a`_has` that references a `_has`.    So for `coverage.payor._has:List:item:_id` it was effectively ignoring the payor part of the chain, skipping to `_has:List:item:_id` when it should have processed `payor._has:List:item:_id` instead, skipping a level of recursion and building incorrect SQL that was missing a JOIN and referencing the wrong resource type.
- Add several new tests for various search strings containing combinations of `_has` and chains.

Closes:  https://github.com/hapifhir/hapi-fhir/issues/5756

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 1]
```diff
@@ -1175,11 +1178,18 @@ private Condition createPredicateHas(
 
 			// Handle internal chain inside the has.
 			if (parameterName.contains(".")) {
-				String chainedPartOfParameter = getChainedPart(parameterName);
+				final String chainedPart =
+						PATTERN_CHAIN_THEN_HAS.matcher(parameterName).find()
+								// The below handles then _has then chain then _has scenario: ex:
+								// "Practitioner?_has:ExplanationOfBenefit:care-team:coverage.payor._has:List:item:_id=list1
+								? getChainedPart(parameterName)
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:1185]: // The below handles the regular **_**has then chain scenario ex:
think ur missing an underscore or i may not understand.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 2]
```diff
@@ -1175,11 +1178,18 @@ private Condition createPredicateHas(
 
 			// Handle internal chain inside the has.
 			if (parameterName.contains(".")) {
-				String chainedPartOfParameter = getChainedPart(parameterName);
+				final String chainedPart =
+						PATTERN_CHAIN_THEN_HAS.matcher(parameterName).find()
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:1182]: oooo big thanks for the comment. there would be no way i could understand without them.

**Review by epeartree — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java [hunk 3]
```diff
@@ -1175,11 +1177,23 @@ private Condition createPredicateHas(
...
 			if (parameterName.contains(".")) {
-				String chainedPartOfParameter = getChainedPart(parameterName);
+				// Previously, for some unknown reason, we were calling getChainedPart() twice.  This broke the _has
+				// then chain, then _has use case by effectively cutting off the second part of the chain and
+				// missing one iteration of the recursive call to build the query.
+				// So, for example, for
+				// Practitioner?_has:ExplanationOfBenefit:care-team:coverage.payor._has:List:item:_id=list1
+				// instead of passing " payor._has:List:item:_id=list1" to the next recursion, the second call to
+				// getChainedPart() was wrongly removing "payor." and passing down "_has:List:item:_id=list1" instead.
+				// This resulted in running incorrect SQL with nonsensical join that resulted in 0 results.
```
Comment (epeartree) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java:1187]: please remove reference to 'cdr'.

**Review by epeartree — CHANGES_REQUESTED** [review]

**Review by epeartree — APPROVED** [review]

**Review by epeartree — APPROVED** [review]
Approved upon changes.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java (hunk 1 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/QueryStack.java]: No longer relevant with the latest fix.

**Pull request merged**